pygsti/algorithms/fiducialselection.py

```diff
@@ -1429,3 +1692,269 @@ def rcl_fn(x): return _scoring.filter_composite_rcl(x, alpha)
+        When `True`, the empty circuit must be a member of the chosen set.
+
+    threshold : float, optional (default is 1e6)
+        Specifies a maximum score for the score matrix, above which the
+        fiducial set is rejected as informationally incomplete.
+
+    verbosity : int, optional
+        How much detail to send to stdout.
+        
+    fid_cache : dict, optional (default is None)
+        A dictionary of either effective state preparations or measurement effects
+        used to accelerate the generation of the matrix used for scoring.
+        It's assumed that the user will pass in the correct cache based on the type
+        of fiducial set being created (if wrong a fall back will revert to redoing all the
+        matrix multiplication again).
+        
+    evd_tol : float, optional (default 1e-10)
+        Cutoff value for truncating small eigenvalues when building low-rank updates
+        caches.
+
+    Returns
+    -------
+    best_fiducials : list
+        The best-scoring list of fiducial circuits.
+
+    """
+    printer = _baseobjs.VerbosityPrinter.create_printer(verbosity)
+
+    if prep_or_meas not in ['prep', 'meas']:
+        raise ValueError("'{}' is an invalid value for prep_or_meas (must be "
+                         "'prep' or 'meas')!".format(prep_or_meas))
+
+    initial_test = test_fiducial_list(model, fids_list, prep_or_meas,
+                                      score_func='all', return_all=False,
+                                      threshold=threshold, fid_cache=fid_cache)
+    if initial_test:
+        printer.log("Complete initial fiducial set succeeds.", 1)
+        printer.log("Now searching for best fiducial set.", 1)
+    else:
+        printer.warning("Complete initial fiducial set FAILS.")
+        printer.warning("Aborting search.")
+        return None, None
+
+    printer.log("Starting fiducial list optimization. Lower score is better.",
+                1)
+    
+    dimRho = model.dim
+    feasibleThreshold = _scoring.CompositeScore(-dimRho, threshold, dimRho)
+    
+    #Build a compact EVD cache for the fiducials
+    fiducial_compact_EVD_cache= construct_compact_evd_cache(model, fids_list, prep_or_meas, fid_cache, eigenvalue_tolerance=evd_tol)
+    
+    #initialize a container for the indices of the candidate fiducial set
+    #along with a bunch of other variables we'll need.
+    best_fiducial_set= None
+    acceptable_candidate_found= False
+    current_best_score_mx= None
+    current_best_score_gramian= None
+    current_update_cache=None
+    current_best_inv_trace= None
+    current_best_composite_score= None
+    
+    #Continue iterating until acceptable solution is found.
+
+    while not acceptable_candidate_found:
+        #if the best fiducial set is not yet initialized
+        #and we have force_empty==True, then we will
+        #initialize it with the index of the empty fiducial.
+        if (best_fiducial_set is None) and force_empty:
```

Comment:
Maybe, for clarity, change this to `if best_fiducial_set is None:` and place nested `if force_empty: ... else: ...` blocks that over the first two cases here.

Reply:
Done.